feat(cardano): add support for alonzo transactions

davidmisiak opened the pull request:
Connect support for https://github.com/vacuumlabs/trezor-firmware/pull/71

davidmisiak committed:
feat(cardano): add key hash stake credentials

davidmisiak commented on src/js/core/methods/CardanoSignTransaction.js:
```diff
@@ -46,6 +49,11 @@ const CardanoSignTransactionFeatures = Object.freeze({
+    OutputDatumHash: ['0', '2.4.4'], // TODO adjust version number
+    ScriptDataHash: ['0', '2.4.4'], // TODO adjust version number
+    Plutus: ['0', '2.4.4'], // TODO adjust version number
+    KeyHashStakeCredential: ['0', '2.4.4'], // TODO adjust version number
```
Are these features ok, or do we want to merge some of them?

davidmisiak commented on src/js/core/methods/CardanoSignTransaction.js:
```diff
@@ -46,6 +49,11 @@ const CardanoSignTransactionFeatures = Object.freeze({
     AuxiliaryDataHash: ['0', '2.4.2'],
     TokenMinting: ['0', '2.4.3'],
     Multisig: ['0', '2.4.3'],
+    NetworkIdInTxBody: ['0', '2.4.4'], // TODO adjust version number
+    OutputDatumHash: ['0', '2.4.4'], // TODO adjust version number
+    ScriptDataHash: ['0', '2.4.4'], // TODO adjust version number
+    Plutus: ['0', '2.4.4'], // TODO adjust version number
+    KeyHashStakeCredential: ['0', '2.4.4'], // TODO adjust version number
```
Are these features ok, or do we want to merge some of them?

gabrielKerekes replied:
I think it's better to have them separately.

davidmisiak commented on tests/__fixtures__/cardanoSignTransaction.js:
```diff
@@ -328,19 +348,41 @@ export default {
             },
+            legacyResults: [
+                {
+                    // networkId not serialized in tx body
```
Do we want legacyResults like this for all test cases? Because that would make the file much longer and less readable, but without them, all the other cases fail when executed with an older firmware (eg. `2.4.2`).

gabrielKerekes replied:
Sadly, I think we should include them, yes.

davidmisiak replied:
Ok, added.

gabrielKerekes commented on docs/methods/cardanoSignTransaction.md:
```diff
@@ -348,14 +483,17 @@ TrezorConnect.cardanoSignTransaction({
 Since transaction streaming has been introduced to the Cardano implementation on Trezor because of memory constraints, Trezor no longer returns the whole serialized transaction as a result of the `CardanoSignTransaction` call. Instead the transaction hash, transaction witnesses and auxiliary data supplement are returned and the serialized transaction needs to be assembled by the client.
 
-###### [flowtype](../../src/js/types/networks/cardano.js#L175-L179)
+Since Alonzo hardfork, `networkId` can be explicitly serialized into the transaction body (which affects the transaction hash and signatures). This provides some security benefits when signing by a hardware wallet, so Trezor will always include it. You can easily find out whether you are deailng with such version of Trezor Firmware by confronting the `isNetworkIdIncludedInTxBody` value.
```
```suggestion
Since Alonzo hardfork, `networkId` can be explicitly serialized into the transaction body (which affects the transaction hash and signatures). This provides some security benefits when signing by a hardware wallet, so Trezor will always include it. You can easily find out whether you are deailng with such version of Trezor Firmware by confronting the value of the `isNetworkIdIncludedInTxBody` field.
```

davidmisiak replied:
Fixed.

gabrielKerekes commented on src/js/types/trezor/protobuf.js:
```diff
@@ -811,6 +818,18 @@ export type CardanoTxMint = {
+// CardanoTxRequiredSigner
+export type CardanoTxRequiredSigner = {
+    key_hash?: string,
+    key_path: number[],
```
I think the `key_path` should be optional. So it needs to be added to `RULE_PATCH` in `scripts/protobuf-patches.js`.
```suggestion
    key_path?: number[],
```

davidmisiak replied:
Fixed. But it's weird that we were sending `CardanoTxRequiredSigner.key_path: undefined` in some cases (see `CardanoSignTransaction.js`) even before the fix and `yarn run flow` didn't complain.

gabrielKerekes replied:
Maybe it's because each part of the `payload.requiredSigners.map` expression is of type `any` and flow just gives up. And maybe it's intentionally setup that way so we wouldn't have to cast every transformation. 🤷 

Actually I just noticed that all the other transformations have definite types. So I think we need to add a cast to the required signers:
```javascript
return ({
  key_path: requiredSigner.keyPath
      ? validatePath(requiredSigner.keyPath, 3)
      : undefined,
  key_hash: requiredSigner.keyHash,
  }: CardanoTxRequiredSigner)
```
When the cast is added, flow does complain about the undefined.
  
And then we should also add a return type to `transformInput` and `transformCollateralInput` functions - sorry I missed it before. In `transformOutput` the return type can be inferred so it doesn't have to be explicitly added, but that apparently isn't the case for the `transformInput` functions.

davidmisiak replied:
:+1: Added.

gabrielKerekes commented on docs/methods/cardanoSignTransaction.md:
```diff
@@ -17,34 +17,39 @@ TrezorConnect.cardanoSignTransaction(params).then(function(result) {
 - *should* have valid `path` property on all `inputs`
 - *must not* contain a pool registration certificate
+- *must not* contain `collateralInputs` and `requiredSigners`
+- *must* contain paths as stake credentials in certificates and withdrawals (no key hases or script hashes)
```
```suggestion
- *must* contain paths as stake credentials in certificates and withdrawals (no key hashes or script hashes)
```

davidmisiak replied:
Fixed.

gabrielKerekes commented on tests/__fixtures__/cardanoSignTransaction.js:
```diff
@@ -18,6 +18,11 @@ const SAMPLE_INPUTS = {
     },
+    shelley_input_2: {
+        path: "m/1852'/1815'/0'/0/1",
+        prev_hash: '3b40265111d8bb3c3c608d95b3a0bf83461ace32d79336579a1939b3aad1c0b7',
```
Maybe `prev_hash` should be changed too.

davidmisiak replied:
Changed.

gabrielKerekes commented on tests/__fixtures__/cardanoSignTransaction.js:
```diff
@@ -250,14 +264,19 @@ const SAMPLE_CERTIFICATES = {
+        amount: '1000',
+    },
+    key_hash: {
+        keyHash: '29fb5fd4aa8cadd6705acc8263cee0fc62edca5ac38db593fec2f9fd',
```
Let's use a `keyHash` which is different from `scriptHash`. E.g. `3a7f09d3df4cf66a7399c2b05bfa234d5a29560c311fc5db4c490711` is used as a keyHash in FW tests. Please also update this in the docs certificates and withdrawals.

davidmisiak replied:
Fixed.

gabrielKerekes commented on tests/__fixtures__/cardanoSignTransaction.js:
```diff
@@ -309,6 +328,7 @@ const SAMPLE_MINTS = {
 const FEE = '42';
 const TTL = '10';
 const VALIDITY_INTERVAL_START = '47';
+const SCRIPT_DATA_HASH = '3b40265111d8bb3c3c608d95b3a0bf83461ace32d79336579a1939b3aad1c0b7';
```
Please use a different hash than is used for inputs.

davidmisiak replied:
Fixed.

gabrielKerekes reviewed (changes requested):
I have just read through the code and I've only tested some basic transactions. I will test it a bit later today. Requested changes are only details.

gabrielKerekes commented:
I have now also tried signing a couple of transactions also on the physical device, I checked the resulting tx body and everything's looking good. 👍  I think this is good to go after my previous comments are fixed.

davidmisiak committed:
feat(cardano): add support for network id in tx body

davidmisiak committed:
feat(cardano): add support for output datum hash and script data hash

davidmisiak committed:
feat(cardano): add support for plutus transactions

davidmisiak committed:
fix(cardano): make input path optional in ts type

gabrielKerekes reviewed (approved):
LGTM except for the `legacyResults` refactor I mentioned offline.

davidmisiak committed:
fix(cardano): update testnet protocol magic

davidmisiak committed:
feat(cardano): allow device-owned outputs in plutus txs

davidmisiak committed:
fix(cardano): fix test legacy results

The .map(fixture => ({ ...fixture, legacyResults })) was overwriting
other legacyResults rules

davidmisiak commented:
Merged in https://github.com/trezor/connect/pull/1073.
